zio-http/src/main/scala/zio/http/Http.scala

```diff
@@ -672,7 +672,8 @@ object Http {
     /**
      * Applies Http based on the path as string
      */
-    def whenPathEq(p: String): HttpApp[R, E] = http.when(_.unsafeEncode.uri().contentEquals(p))
+    def whenPathEq(p: String): HttpApp[R, E] =
+      http.when(request => Unsafe.unsafe { implicit u => request.unsafe.encode.uri().contentEquals(p) })
```

Comment:
Unsafety not handled

Reply:
Ditto.

Reply:
This will result in performance degradation. 
The older implementation was able to match path with Zero allocations.

Reply:
@tusharmath I think the user can opt out of this by providing an implicit `Unsafe` once (by getting and storing one initialization time because the singleton value is `private[zio]`) and then there is no extra allocations compared to the original version

Reply:
@tusharmath I think with @jdegoes's recommended change the user can opt out of this by providing an implicit `Unsafe` once (by getting and storing one initialization time because the singleton value is `private[zio]`) and then there is no extra allocations compared to the original version. (and even without that it is not per request but at `HttpApp` definition time)

Reply:
@tusharmath Yes, there would have been performance degradation with `Unsafe.unsafe { ... }`, but if we reflect that in the API with (implicit unsafe) rather than trying to "escape" the unsafeness, then there will be no impact on performance, since ZIO HTTP will itself simply import the unsafe implicit into scope.

Reply:
There is still an additional `Function` that's being allocated. Which is degrading the performance.

Earlier we used to be at around 1.01M rps now its around 0.80 rps.